src/utils.js

```diff
@@ -528,37 +527,67 @@ const isObjEmpty = (obj) => {
   return true;
 };
 
+/** Recursively attach comments to nodes */
+function attachCommentsHTML(node) {
+  if (!isNodeWithChildren(node) || !node.children.some(({ type }) => type === 'Comment')) return;
+
+  const nodesToRemove = [];
+
+  // note: the .length - 1 is because we don’t need to read the last node
+  for (let n = 0; n < node.children.length - 1; n++) {
+    if (!node.children[n]) continue;
+
+    // attach comment to the next non-whitespace node
+    if (node.children[n].type === 'Comment') {
+      let next = n + 1;
+      while (isEmptyTextNode(node.children[next])) {
+        nodesToRemove.push(next); // if arbitrary whitespace between comment and node, remove
+        next++; // skip to the next non-whitespace node
+      }
+      util.addLeadingComment(node.children[next], node.children[n]);
+    }
+  }
+
+  // remove arbitrary whitespace nodes
```

Comment:
I found this behavior necessary because Prettier kept trying to preserve the whitespace between the actual Prettier comment and the node it was attached to. But despite moving whitespace around, Prettier seemed to rearrange everything correctly in the following pass when it [processes the comment](https://prettier.io/docs/en/plugins.html#handling-comments-in-a-printer)